lib/models/package-info-cache.js

```diff
@@ -0,0 +1,1049 @@
+      // a node_modules inside it, so check there first because it's
+      // quicker since we have the reference. Only check externally
+      // if we don't find it there.
+      if (this.nodeModules) {
+        dependencyPackage = this.nodeModules._findPackage(dependencyName);
+      }
+
+      if (!dependencyPackage) {
+        dependencyPackage = this.cache._findPackage(
+          dependencyName,
+          path.dirname(this.realPath)
+        );
+      }
+
+      if (dependencyPackage) {
+        packages[dependencyName] = dependencyPackage;
+      } else {
+        missingDependencies.push(dependencyName);
+      }
+    });
+
+    if (missingDependencies.length > 0) {
+      this._addError(ERROR_DEPENDENCIES_MISSING, missingDependencies);
+    }
+
+    return packages;
+  }
+
+  isAddon() {
+    let val =
+      Array.isArray(this.pkg.keywords) &&
+      this.pkg.keywords.indexOf('ember-addon') >= 0;
+    return val;
+  }
+
+  // XXX Should this move to Addon.js (and change appropriately to use the cache here?)
+  getAddonConstructor() {
+    if (this.isAddon() && !this.addonConstructor) {
+      if (!this.addonMainPath) {
+        return null;
+      }
+
+      let Addon = require('./addon');
+
+      // Load the addon.
+      // TODO: Future work - allow a time budget for loading each addon and warn
+      // or error for those that take too long.
+      let module = require(this.addonMainPath);
+      let mainDir = path.dirname(this.addonMainPath);
+
+      let ctor;
+
+      if (typeof module === 'function') {
+        ctor = module;
+        ctor.prototype.root = ctor.prototype.root || mainDir;
+        ctor.prototype.pkg = ctor.prototype.pkg || this.pkg;
+      } else {
+        ctor = Addon.extend(
+          Object.assign({ root: mainDir, pkg: this.pkg }, module)
+        );
+      }
+
+      ctor._meta_ = {
+        modulePath: this.addonMainPath,
+        //        lookupDuration,
+        initializeIn: 0,
+      };
+
+      this.addonConstructor = ctor;
+    }
+
+    return this.addonConstructor;
+  }
+}
+
+class NodeModulesList {
+  constructor(realPath, cache) {
+    this.realPath = realPath;
+    this.cache = cache;
+    this.entries = Object.create(null);
+    this.errors = new ErrorList();
+  }
+
+  _addError(errorType, errorData) {
+    this.errors.addError(errorType, errorData);
+  }
+
+  _insertError(errorEntry) {
+    this.errors.insertError(errorEntry);
+  }
+
+  _hasErrors() {
+    return this.errors.hasErrors();
+  }
+
+  _addEntry(entryName, entryVal) {
+    this.entries[entryName] = entryVal;
+  }
+
+  /**
+   * Return a PackageInfo object for a given package name (which may include
+   * a scope)
+   *
+   * @private
+   * @method _findPackage
+   * @param {String} packageName the name (possibly including a scope) of
+   *    the PackageInfo the caller wants returned.
+   * @return the desired PackageInfo if one exists for the given name, else null.
+   */
+  _findPackage(packageName) {
```

Comment:
unit test

Reply:
A question: are we supposed to be able to unit test internal functions like this one directly, or via side-effect, or what? _findPackage (in NodeModulesList) is only called as part of _addDependencies, and that would